File: examples/csharp/PrjOutput/TextRenderer.cs
using Google.Protobuf;
using Angzarr.Examples;

namespace PrjOutput;

/// <summary>
/// Renders events as human-readable text.
/// </summary>
public class TextRenderer
{
    private readonly Dictionary<string, string> _playerNames = new();

    public void SetPlayerName(ByteString playerRoot, string name)
    {
        var key = Convert.ToHexString(playerRoot.ToByteArray()).ToLowerInvariant();
        _playerNames[key] = name;
    }

    public string GetPlayerName(ByteString playerRoot)
    {
        var key = Convert.ToHexString(playerRoot.ToByteArray()).ToLowerInvariant();
        return _playerNames.TryGetValue(key, out var name) ? name : key[..8];
    }

    public string Render(string eventType, object evt)
    {
        return eventType switch
        {
            "PlayerRegistered" => RenderPlayerRegistered((PlayerRegistered)evt),
            "FundsDeposited" => RenderFundsDeposited((FundsDeposited)evt),
            "FundsWithdrawn" => RenderFundsWithdrawn((FundsWithdrawn)evt),
            "FundsReserved" => RenderFundsReserved((FundsReserved)evt),
            "FundsReleased" => RenderFundsReleased((FundsReleased)evt),
            "TableCreated" => RenderTableCreated((TableCreated)evt),
            "PlayerJoined" => RenderPlayerJoined((PlayerJoined)evt),
            "PlayerLeft" => RenderPlayerLeft((PlayerLeft)evt),
            "HandStarted" => RenderHandStarted((HandStarted)evt),
            "HandEnded" => RenderHandEnded((HandEnded)evt),
            "CardsDealt" => RenderCardsDealt((CardsDealt)evt),
            "BlindPosted" => RenderBlindPosted((BlindPosted)evt),
            "ActionTaken" => RenderActionTaken((ActionTaken)evt),
            "CommunityCardsDealt" => RenderCommunityCardsDealt((CommunityCardsDealt)evt),
            "PotAwarded" => RenderPotAwarded((PotAwarded)evt),
            "HandComplete" => RenderHandComplete((HandComplete)evt),
            _ => $"[{eventType}]"
        };
    }

    private string RenderPlayerRegistered(PlayerRegistered evt)
    {
        return $"Player '{evt.DisplayName}' registered ({evt.PlayerType})";
    }

    private string RenderFundsDeposited(FundsDeposited evt)
    {
        return $"Deposited {evt.Amount?.Amount} chips (balance: {evt.NewBalance?.Amount})";
    }

    private string RenderFundsWithdrawn(FundsWithdrawn evt)
    {
        return $"Withdrew {evt.Amount?.Amount} chips (balance: {evt.NewBalance?.Amount})";
    }

    private string RenderFundsReserved(FundsReserved evt)
    {
        return $"Reserved {evt.Amount?.Amount} chips for table";
    }

    private string RenderFundsReleased(FundsReleased evt)
    {
        return $"Released {evt.Amount?.Amount} chips from table";
    }

    private string RenderTableCreated(TableCreated evt)
    {
        return $"Table '{evt.TableName}' created ({evt.GameVariant}, {evt.SmallBlind}/{evt.BigBlind})";
    }

    private string RenderPlayerJoined(PlayerJoined evt)
    {
        var player = GetPlayerName(evt.PlayerRoot);
        return $"{player} joined seat {evt.SeatPosition} with {evt.BuyInAmount} chips";
    }

    private string RenderPlayerLeft(PlayerLeft evt)
    {
        var player = GetPlayerName(evt.PlayerRoot);
        return $"{player} left with {evt.ChipsCashedOut} chips";
    }

    private string RenderHandStarted(HandStarted evt)
    {
        return $"=== Hand #{evt.HandNumber} Started (Dealer: seat {evt.DealerPosition}) ===";
    }

    private string RenderHandEnded(HandEnded evt)
    {
        return $"=== Hand Ended ===";
    }

    private string RenderCardsDealt(CardsDealt evt)
    {
        return $"Cards dealt to {evt.PlayerCards.Count} players";
    }

    private string RenderBlindPosted(BlindPosted evt)
    {
        var player = GetPlayerName(evt.PlayerRoot);
        return $"{player} posts {evt.BlindType} blind: {evt.Amount}";
    }

    private string RenderActionTaken(ActionTaken evt)
    {
        var player = GetPlayerName(evt.PlayerRoot);
        var action = evt.Action switch
        {
            ActionType.Fold => "folds",
            ActionType.Check => "checks",
            ActionType.Call => $"calls {evt.Amount}",
            ActionType.Bet => $"bets {evt.Amount}",
            ActionType.Raise => $"raises to {evt.Amount}",
            ActionType.AllIn => $"all-in for {evt.Amount}",
            _ => evt.Action.ToString()
        };
        return $"{player} {action} (pot: {evt.PotTotal})";
    }

    private string RenderCommunityCardsDealt(CommunityCardsDealt evt)
    {
        var cards = string.Join(" ", evt.Cards.Select(RenderCard));
        var phase = evt.Phase.ToString().ToUpper();
        return $"*** {phase}: {cards} ***";
    }

    private string RenderPotAwarded(PotAwarded evt)
    {
        var winners = evt.Winners.Select(w =>
        {
            var player = GetPlayerName(w.PlayerRoot);
            return $"{player} wins {w.Amount}";
        });
        return string.Join(", ", winners);
    }

    private string RenderHandComplete(HandComplete evt)
    {
        return $"Hand #{evt.HandNumber} complete";
    }

    private static string RenderCard(Card card)
    {
        var rank = card.Rank switch
        {
            Rank.Two => "2",
            Rank.Three => "3",
            Rank.Four => "4",
            Rank.Five => "5",
            Rank.Six => "6",
            Rank.Seven => "7",
            Rank.Eight => "8",
            Rank.Nine => "9",
            Rank.Ten => "T",
            Rank.Jack => "J",
            Rank.Queen => "Q",
            Rank.King => "K",
            Rank.Ace => "A",
            _ => "?"
        };
        var suit = card.Suit switch
        {
            Suit.Clubs => "c",
            Suit.Diamonds => "d",
            Suit.Hearts => "h",
            Suit.Spades => "s",
            _ => "?"
        };
        return rank + suit;
    }
}

File: examples/cpp/prj-output/src/text_renderer.cpp
#include "text_renderer.hpp"
#include <sstream>
#include <iomanip>

namespace projector {

void TextRenderer::set_player_name(const std::string& player_root, const std::string& name) {
    // Convert bytes to hex for key
    std::stringstream ss;
    for (unsigned char c : player_root) {
        ss << std::hex << std::setfill('0') << std::setw(2) << static_cast<int>(c);
    }
    player_names_[ss.str()] = name;
}

std::string TextRenderer::get_player_name(const std::string& player_root) const {
    std::stringstream ss;
    for (unsigned char c : player_root) {
        ss << std::hex << std::setfill('0') << std::setw(2) << static_cast<int>(c);
    }
    std::string key = ss.str();

    auto it = player_names_.find(key);
    if (it != player_names_.end()) {
        return it->second;
    }
    // Return shortened hex if no name set
    return key.length() >= 8 ? key.substr(0, 8) : key;
}

std::string TextRenderer::render_card(const examples::Card& card) {
    std::string rank;
    if (card.rank() == 14) rank = "A";
    else if (card.rank() == 13) rank = "K";
    else if (card.rank() == 12) rank = "Q";
    else if (card.rank() == 11) rank = "J";
    else if (card.rank() == 10) rank = "T";
    else rank = std::to_string(card.rank());

    std::string suit;
    switch (card.suit()) {
        case examples::SPADES: suit = "♠"; break;
        case examples::HEARTS: suit = "♥"; break;
        case examples::DIAMONDS: suit = "♦"; break;
        case examples::CLUBS: suit = "♣"; break;
        default: suit = "?"; break;
    }

    return rank + suit;
}

std::string TextRenderer::render_action(examples::ActionType action) {
    switch (action) {
        case examples::FOLD: return "folds";
        case examples::CHECK: return "checks";
        case examples::CALL: return "calls";
        case examples::BET: return "bets";
        case examples::RAISE: return "raises";
        case examples::ALL_IN: return "all-in";
        default: return "unknown";
    }
}

std::string TextRenderer::render_player_registered(const examples::PlayerRegistered& event) {
    std::stringstream ss;
    ss << "Player '" << event.display_name() << "' registered";
    if (event.player_type() == examples::AI) {
        ss << " (AI)";
    }
    return ss.str();
}

std::string TextRenderer::render_funds_deposited(const examples::FundsDeposited& event) {
    std::stringstream ss;
    ss << "Deposited " << event.amount().amount()
       << " (new balance: " << event.new_balance().amount() << ")";
    return ss.str();
}

std::string TextRenderer::render_funds_withdrawn(const examples::FundsWithdrawn& event) {
    std::stringstream ss;
    ss << "Withdrew " << event.amount().amount()
       << " (new balance: " << event.new_balance().amount() << ")";
    return ss.str();
}

std::string TextRenderer::render_funds_reserved(const examples::FundsReserved& event) {
    std::stringstream ss;
    ss << "Reserved " << event.amount().amount() << " for table";
    return ss.str();
}

std::string TextRenderer::render_funds_released(const examples::FundsReleased& event) {
    std::stringstream ss;
    ss << "Released reserved funds (new balance: " << event.new_available_balance().amount() << ")";
    return ss.str();
}

std::string TextRenderer::render_table_created(const examples::TableCreated& event) {
    std::stringstream ss;
    ss << "Table '" << event.table_name() << "' created - "
       << event.small_blind() << "/" << event.big_blind() << " blinds, "
       << "max " << event.max_players() << " players";
    return ss.str();
}

std::string TextRenderer::render_player_joined(const examples::PlayerJoined& event) {
    std::stringstream ss;
    ss << get_player_name(event.player_root()) << " joined at seat "
       << event.seat_position() << " with " << event.stack();
    return ss.str();
}

std::string TextRenderer::render_player_left(const examples::PlayerLeft& event) {
    std::stringstream ss;
    ss << get_player_name(event.player_root()) << " left with "
       << event.chips_cashed_out();
    return ss.str();
}

std::string TextRenderer::render_hand_started(const examples::HandStarted& event) {
    std::stringstream ss;
    ss << "=== Hand #" << event.hand_number() << " ===\n"
       << "Dealer: seat " << event.dealer_position() << ", "
       << "Blinds: " << event.small_blind() << "/" << event.big_blind();
    return ss.str();
}

std::string TextRenderer::render_hand_ended(const examples::HandEnded& event) {
    return "Hand ended";
}

std::string TextRenderer::render_cards_dealt(const examples::CardsDealt& event) {
    std::stringstream ss;
    ss << "Cards dealt to " << event.player_cards_size() << " players";
    return ss.str();
}

std::string TextRenderer::render_blind_posted(const examples::BlindPosted& event) {
    std::stringstream ss;
    ss << get_player_name(event.player_root()) << " posts "
       << event.blind_type() << " blind: " << event.amount();
    return ss.str();
}

std::string TextRenderer::render_action_taken(const examples::ActionTaken& event) {
    std::stringstream ss;
    ss << get_player_name(event.player_root()) << " "
       << render_action(event.action());
    if (event.amount() > 0) {
        ss << " " << event.amount();
    }
    return ss.str();
}

std::string TextRenderer::render_community_cards_dealt(const examples::CommunityCardsDealt& event) {
    std::stringstream ss;
    ss << "*** ";
    switch (event.phase()) {
        case examples::FLOP: ss << "FLOP"; break;
        case examples::TURN: ss << "TURN"; break;
        case examples::RIVER: ss << "RIVER"; break;
        default: ss << "COMMUNITY"; break;
    }
    ss << " *** [";
    bool first = true;
    for (const auto& card : event.all_community_cards()) {
        if (!first) ss << " ";
        ss << render_card(card);
        first = false;
    }
    ss << "]";
    return ss.str();
}

std::string TextRenderer::render_pot_awarded(const examples::PotAwarded& event) {
    std::stringstream ss;
    ss << "*** POT AWARDED ***\n";
    for (const auto& winner : event.winners()) {
        ss << get_player_name(winner.player_root()) << " wins "
           << winner.amount() << "\n";
    }
    return ss.str();
}

std::string TextRenderer::render_hand_complete(const examples::HandComplete& event) {
    std::stringstream ss;
    ss << "=== Hand Complete ===\n";
    ss << "Final stacks:\n";
    for (const auto& stack : event.final_stacks()) {
        ss << "  " << get_player_name(stack.player_root()) << ": "
           << stack.stack() << "\n";
    }
    return ss.str();
}

} // namespace projector
